### PR TITLE
Drop-down filtering

### LINKED ISSUE #9058: Drop-down widget filtering
* The drop-down should be shown on any click on the widget, not just the arrow.
* When the internal widget of the drop-down is something editable (text or numeric input), _once user will start typing_, the options list should be filtered by the current content.

### PR DESCRIPTION
**Pull request by farmaazon**
### Pull Request Description

Fixes #9058 - the filtering so far is a bit aggressive, but I tune it up in next PR(s).

[Screencast from 2024-03-13 15-20-17.webm](https://github.com/enso-org/enso/assets/3919101/112ce65a-a8c6-4818-b8b8-9f493caf9c81)

Added new special `WidgetEditHandler,` allowing handling "multi-widget" interactions needed for drop down filtering.

### Important Notes

* Now when clicking on argument name, the edit is accepted (as normal "outside" click), and then the dropdown is opened again (due to handling click event). I didn't figure out how to handle this case properly, left something least confusing.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [ ] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Comment by farmaazon** [comment]
The logic is ready for review, I will yet add tests to undraft the PR.

File: app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue [hunk 1]
```diff
@@ -62,6 +66,19 @@ function tagFromEntry(entry: SuggestionEntry): Tag {
   }
 }
 
+function isFilteredIn(tag: Tag): boolean {
+  const pattern =
+    editedValue.value instanceof Ast.Ast ? editedValue.value.code() : editedValue.value
+  if (!pattern) return true
+  const textLiteralPattern = /^(['"])(.*)\1$/.exec(pattern)
```
Comment (vitvakatu) [inline app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue]: I guess it’s better to parse with the parser. String literals can also start with triple quotes.

File: app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue [hunk 2]
```diff
@@ -62,6 +66,19 @@ function tagFromEntry(entry: SuggestionEntry): Tag {
   }
 }
 
+function isFilteredIn(tag: Tag): boolean {
+  const pattern =
```
Comment (kazcw) [inline app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue]: Could extract `tag`-invariant computation

File: app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue [hunk 3]
```diff
@@ -112,29 +131,53 @@ const selectedLabel = computed(() => {
...
+    ...props.input,
+    editHandler: dropDownInteraction,
+  }
+  if (props.input.dynamicConfig == null) return newInput
   const config = props.input.dynamicConfig
-  if (config.kind !== 'Single_Choice') return props.input
-  return { ...props.input, dynamicConfig: singleChoiceConfiguration(config) }
+  if (config.kind !== 'Single_Choice') return newInput
+  newInput.dynamicConfig = singleChoiceConfiguration(config)
+  return newInput
```
Comment (kazcw) [inline app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue]: Could simplify this now and avoid early-exits

File: app/gui2/src/components/GraphEditor/widgets/WidgetText.vue [hunk 4]
```diff
@@ -1,14 +1,49 @@
...
+function accepted() {
+  editing.end()
+  if (props.input.value instanceof Ast.TextLiteral) {
+    const edit = graph.startEdit()
+    edit.getVersion(props.input.value).setRawTextContent(editedContents.value)
+    props.onUpdate({ edit })
+  } else {
+    props.onUpdate({
+      portUpdate: {
+        value: makeNewLiteral(editedContents.value).code(),
```
Comment (kazcw) [inline app/gui2/src/components/GraphEditor/widgets/WidgetText.vue]: Why stringify here?

File: app/gui2/src/components/GraphEditor/widgets/WidgetText.vue [hunk 5]
```diff
@@ -21,29 +56,24 @@ function makeNewLiteral(value: string) {
   return Ast.TextLiteral.new(value, MutableModule.Transient())
 }
 
+function makeLiteralFromValue(value: string) {
```
Comment (kazcw) [inline app/gui2/src/components/GraphEditor/widgets/WidgetText.vue]: Confusing function signature

File: app/gui2/src/providers/widgetRegistry/__tests__/editHandler..test.ts [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    expect(editedHandler.handler.isActive()).toBeFalsy()
+  },
+)
+
+// ${'Propagating'}                    | ${['A', 'A1']}        | ${'A1'}  | ${['A', 'A1']}      | ${[]}                  | ${['A', 'A1']}
+// ${'Parent edited'}                  | ${['A', 'A1']}        | ${'A'}   | ${['A', 'A1']}      | ${[]}                  | ${['A']}
+// ${'Not propagating'}                | ${['A', 'A1']}        | ${'A1'}  | ${['A1']}           | ${['A']}               | ${['A']}
+
+// ${'Skipping handler without click'} | ${['A', 'A1', 'A12']} | ${'A12'} | ${['A', 'A12']}     | ${[]}                  | ${['A', 'A12']}
+//   ${'Stopping propagation'}           | ${['A', 'A1', 'A12']} | ${'A12'} | ${['A', 'A12']}     | ${['A1']}              | ${['A', 'A1']}
```
Comment (kazcw) [inline app/gui2/src/providers/widgetRegistry/__tests__/editHandler..test.ts]: Commented cases?

File: app/gui2/src/providers/widgetRegistry/__tests__/editHandler..test.ts [hunk 7]
```diff
@@ -0,0 +1,147 @@
+import type { GraphNavigator } from '@/providers/graphNavigator'
```
Comment (kazcw) [inline app/gui2/src/providers/widgetRegistry/__tests__/editHandler..test.ts]: Extra '.' in name of this file

File: app/gui2/src/providers/widgetRegistry/editHandler.ts [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+  start?(origin: PortId): void
+  edit?(origin: PortId, value: Ast.Owned | string): void
+  end?(origin: PortId): void
+}
+
+/**
+ * Widget edit handler.
+ *
+ * This handler takes an extended interaction and allows cooperation between parent/child
+ * interactions. A usage example is WidgetSelection, which want's to open when the child is edited
```
Comment (vitvakatu) [inline app/gui2/src/providers/widgetRegistry/editHandler.ts]:
```suggestion
 * interactions. A usage example is WidgetSelection, which wants to open when the child is edited
```

File: app/gui2/src/providers/widgetRegistry/editHandler.ts [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+ * and filters entries by edited temporary value.
+ *
+ * Widget's edit state should be manipulated by `start`, `end` and `cancel` methods; they will set
+ * proper interaction in the global {@link InteractionHandler} and calls the additional callbacks in
```
Comment (vitvakatu) [inline app/gui2/src/providers/widgetRegistry/editHandler.ts]:
```suggestion
 * proper interaction in the global {@link InteractionHandler} and call the additional callbacks in
```

File: app/gui2/src/providers/widgetRegistry/editHandler.ts [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+ * {@link WidgetEditInteraction} passed during construction.
+ *
+ * The parent widget may pass its edit handler to one or more children's {@link WidgetInput} to
+ * bound their interactions; when this child is edited, the parent is also considered edited,
+ * along with any further ancestors. In particular:
+ * - Starting, ending and cancelling (including automatic canceling by the global interaction
+ *   handler) of child edit will also call proper callbacks in parent.
+ * - Cancelling or ending parent edit will candel/end the child's interaction.
```
Comment (vitvakatu) [inline app/gui2/src/providers/widgetRegistry/editHandler.ts]:
```suggestion
 * - Cancelling or ending parent edit will cancel/end the child's interaction.
```

File: app/gui2/src/providers/widgetRegistry/editHandler.ts [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+  injectInteractionHandler,
+  type Interaction,
+  type InteractionHandler,
+} from '@/providers/interactionHandler'
+import type { WidgetInput } from '@/providers/widgetRegistry'
+import type { Ast } from '@/util/ast'
+
+/** An extend {@link Interaction} used in {@link WidgetEditHandler} */
+export interface WidgetEditInteraction extends Interaction {
+  /** Click handler from {@link Interaction}, but received child click handler. See
```
Comment (vitvakatu) [inline app/gui2/src/providers/widgetRegistry/editHandler.ts]: I don’t understands the second part of the sentence, after the `but`.

File: app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue [hunk 12]
```diff
@@ -122,39 +177,65 @@ const selectedTag = computed(() => {
...
 })
-const innerWidgetInput = computed(() => {
-  if (props.input.dynamicConfig == null) return props.input
-  const config = props.input.dynamicConfig
-  if (config.kind !== 'Single_Choice') return props.input
-  return { ...props.input, dynamicConfig: singleChoiceConfiguration(config) }
+const innerWidgetInput = computed<WidgetInput>(() => {
+  const dynamicConfig =
+    props.input.dynamicConfig?.kind === 'Single_Choice' ?
+      singleChoiceConfiguration(props.input.dynamicConfig)
```
Comment (vitvakatu) [inline app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue:187]: Before, we were passing down the dynamic config if it is *not* Single_Choice. Now we don’t. Is it deliberate?

File: app/gui2/src/components/GraphEditor/widgets/WidgetText.vue [hunk 13]
```diff
@@ -60,9 +89,19 @@ export const widgetDefinition = defineWidget(WidgetInput.isAstOrPlaceholder, {
...
     <NodeWidget v-if="shownLiteral.open" :input="WidgetInput.FromAst(shownLiteral.open)" />
-    <AutoSizedInput v-model.lazy="textContents" />
+    <AutoSizedInput
+      ref="input"
+      v-model="editedContents"
+      autoSelect
+      @pointerdown.stop
+      @pointerup.stop
+      @click.stop
+      @keydown.enter.stop="accepted"
```
Comment (vitvakatu) [inline app/gui2/src/components/GraphEditor/widgets/WidgetText.vue:101]: Not sure, but `escape` can probably cancel?

File: app/gui2/src/providers/widgetRegistry/editHandler.ts [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+ * will receive `cancel` feedback from the latter and then `start` from the former.
+ *
+ * **The `click` handler is a special case:** it will be called only on top-most parent, but its
+ * handler may decide to delegate it further by calling child's handler passed as an additional
+ * argument
+ */
+export class WidgetEditHandler {
+  private interaction: WidgetEditInteraction
+  /** This, or one's child interaction which is currently active */
+  private currentInteraction: WidgetEditInteraction | undefined
```
Comment (vitvakatu) [inline app/gui2/src/providers/widgetRegistry/editHandler.ts]: The naming for these two fields is a bit confusing, but I guess it’s okay.

File: app/gui2/src/components/GraphEditor/widgets/WidgetText.vue (hunk 13 above)
Comment (farmaazon) [reply app/gui2/src/components/GraphEditor/widgets/WidgetText.vue:101]: It cancels already, but through our interaction system. Each active interaction is cancelled by Esc (caught in capture phase on window level).

File: app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue (hunk 12 above)
Comment (farmaazon) [reply app/gui2/src/components/GraphEditor/widgets/WidgetSelection.vue:187]: No. I restored the old way, although I think there is not much difference.

**Review by vitvakatu — APPROVED** [review]

**Pull request merged**